### ISSUE TITLE
Incorrect UserWarning for dataloader edge case

### ISSUE THREAD

**Issue by jrichardson97** [PyTorchLightning/pytorch-lightning#10790]
Using pytorch-lightning 1.5.2 and and pytorch 1.10.0. 

I'm using a custom DataLoader initialization that sends a list of indices to the __getitem__ method of my datasets to support single worker slicing of an in memory, large dataset ([discussed here](https://discuss.pytorch.org/t/dataloader-sample-by-slices-from-dataset/113005/5)). To do this, you pass a BatchSampler as the sampler. Additionally, I am passing batch_size=None to avoid the default collate function adding an additional dimension and manually set shuffle=False. I've verified everything works as expected with pytorch lightning through the debugger.

The following check in pytorch_lightning.trainer.data_loading.TrainerDataLoadingMixin._reset_eval_dataloader gives me a warning that is incorrect for this use case:
```python
if hasattr(loader, "sampler") and not isinstance(loader.sampler, SequentialSampler):
    # when overfitting, the dataloader should not have sampler
    if self.overfit_batches > 0 and mode.evaluating:
        rank_zero_warn(
            "You requested to overfit but enabled val/test dataloader shuffling."
            " We are turning it off for you."
        )
        dataloaders[loader_i] = self._update_dataloader(
            loader, SequentialSampler(loader.dataset), mode=mode
        )
    else:
        rank_zero_warn(
            f"Your `{mode.dataloader_prefix}_dataloader` has `shuffle=True`,"
            "it is strongly recommended that you turn this off for val/test/predict dataloaders."
        )
```

My dataloaders do have a sampler, but they are BatchSamplers that each contain a SequentialSampler. I think the check was written this way since shuffle isn't saved as an attribute during DataLoader initialization. An additional check for 'isinstance(loader.sampler.sampler, SequentialSampler)' would eliminate the warning for my case, but I'm not sure if this would cause issues elsewhere.
